common/nymcoconut/src/proofs/mod.rs

```diff
@@ -0,0 +1,691 @@
+            &response_private_elgamal_key_bytes,
+            CoconutError::Deserialization(
+                "Failed to deserialize the response to the private ElGamal key".to_string(),
+            ),
+        )?;
+
+        let rk_len = u64::from_le_bytes(bytes[idx..idx + 8].try_into().unwrap());
+        idx += 8;
+        if bytes[idx..].len() < rk_len as usize * 32 + 8 {
+            return Err(
+                CoconutError::Deserialization(
+                    "tried to deserialize proof of ciphertexts and commitment with insufficient number of bytes provided".to_string()),
+            );
+        }
+
+        let rk_end = idx + rk_len as usize * 32;
+        let response_keys = try_deserialize_scalar_vec(
+            rk_len,
+            &bytes[idx..rk_end],
+            CoconutError::Deserialization("Failed to deserialize keys response".to_string()),
+        )?;
+
+        let rm_len = u64::from_le_bytes(bytes[rk_end..rk_end + 8].try_into().unwrap());
+        let response_attributes = try_deserialize_scalar_vec(
+            rm_len,
+            &bytes[rk_end + 8..],
+            CoconutError::Deserialization("Failed to deserialize attributes response".to_string()),
+        )?;
+
+        Ok(ProofCmCs {
+            challenge,
+            response_opening,
+            response_private_elgamal_key,
+            response_keys,
+            response_attributes,
+        })
+    }
+}
+
+#[derive(Debug)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct ProofKappaNu {
+    // c
+    challenge: Scalar,
+
+    // responses
+    response_serial_number: Scalar,
+    response_binding_number: Scalar,
+    response_blinder: Scalar,
+}
+
+impl ProofKappaNu {
+    pub(crate) fn construct(
+        params: &Parameters,
+        verification_key: &VerificationKey,
+        serial_number: &Attribute,
+        binding_number: &Attribute,
+        blinding_factor: &Scalar,
+        blinded_message: &G2Projective,
+        blinded_serial_number: &G2Projective,
+    ) -> Self {
+        // create the witnesses
+        let witness_blinder = params.random_scalar();
+        let witness_serial_number = params.random_scalar();
+        let witness_binding_number = params.random_scalar();
+        let witness_attributes = vec![witness_serial_number, witness_binding_number];
+
+        let beta_bytes = verification_key
+            .beta
+            .iter()
+            .map(|beta_i| beta_i.to_bytes())
+            .collect::<Vec<_>>();
+
+        // witnesses commitments
+        // Aw = g2 * wt + alpha + beta[0] * wm[0] + ... + beta[i] * wm[i]
+        let commitment_kappa = params.gen2() * witness_blinder
+            + verification_key.alpha
+            + witness_attributes
+                .iter()
+                .zip(verification_key.beta.iter())
+                .map(|(wm_i, beta_i)| beta_i * wm_i)
+                .sum::<G2Projective>();
+
+        // zeta is the public value associated with the serial number
```

Comment:
So maybe we could call it SerialNumber instead of zeta? : )

Reply:
zeta is not a serial number, it is g^serial_number. At some point, I will try to rethink how to name all those things using English :)